### PR TITLE
Add support for seed in `DataCollatorForLanguageModeling`

### LINKED ISSUE #36357: Allow setting a seed for DataCollatorForLanguageModeling
### Feature request

The `DataCollatorForLanguageModeling` class allows training for an MLM (masked language model) task, which randomly masks or replaces certain tokens. Models such as BERT and RoBERTa are trained in such a manner. It would be great if the user can set a seed, ensuring repeatability in generating masked batches.

### Motivation

This would ensure generation of repeatable batches of data, which is critical for model reproducibility. Right now, there is a form of repeatability with `transformers.set_seed()`, but one can make use of generators([PyTorch](https://pytorch.org/docs/stable/generated/torch.Generator.html), [Tensorflow](https://www.tensorflow.org/api_docs/python/tf/random/Generator), [NumPy](https://numpy.org/doc/2.1/reference/random/generator.html)) to set the data collator seed without globally setting the seed for each framework. The major reason this would help is that the MLM masking probabilities would not be influenced by code outside of it, which is good practice. This would mean that, given the same dataset and seed, the masking would happen consistently irrespective of the rest of your training script. See [this blog post for more details](https://albertcthomas.github.io/good-practices-random-number-generators/#random-number-generation-with-numpy). 

### Your contribution

I can submit a PR for this. I have experience with TF, PyTorch and NumPy, would love to contribute. I have taken a look at the code, and can add a `seed` argument which enables usage of generators for repeatability. If not specified, however, the code would fall back to its previous behavior, including using `transformers.set_seed()`.

### PR DESCRIPTION
**Pull request by capemox**
# What does this PR do?
This PR adds support for setting a seed in the class `DataCollatorForLanguageModeling`. This helps with reproducibility in generating masks for masked language modeling (MLM). This issue was approved by @Rocketknight1 (https://github.com/huggingface/transformers/issues/36357)

Currently, there is a way for ensuring reproducibility by using the function `transformers.set_seed()`. However, this function sets the seed of the global RNG for PyTorch, Numpy, etc. What this means is that, setting a global seed can impact other pseudo-random functions outside the scope of the collator such as parameter initialization for models. This also means, that changes in the script outside the collator can impact the masking.

Instead, it is preferred to create generator objects which can be passed around to different functions. [This is also considered good practice](https://blog.scientific-python.org/numpy/numpy-rng/). What my PR does, is:

- Allows users to pass a `seed` parameter to `DataCollatorForLanguageModeling`
- Instantiates a generator object with the seed depending on the `return_tensors` parameter
- The generator object is used for pseudo-random functions within the class
- In case the user does not pass a seed, the collator falls back to its default implementation, for backwards compatibility

The generator object is scoped to the collator class, so it won't affect pseudo-random functions outside the class and vice-versa.

One important factor to consider is using multiple workers for the collator function, as PyTorch's `DataLoader` does. [PyTorch has documentation regarding this](https://pytorch.org/docs/stable/data.html#randomness-in-multi-process-data-loading), whereby we set a different seed for each worker given by `shared_seed + worker_id`. This is because the worker's seeds would be cloned, and so each worker would mask the input in exactly the same manner, which is undesirable. A critical part of PyTorch's `DataLoader` is that from within the worker, it is possible to access the worker's `id` (important to set the worker seed). Because of this constraint, this PR **only supports multi-worker scenarios with PyTorch's `DataLoader`**. With the `seed` set, if the code detects that the collator is running in a multi-processing scenario and the worker information is unavailable, an error is thrown.

The algorithm for creating the generator object is:
```python
# use multiprocessing to get current process' name
if current_process == "MainProcess":
    # Collator running in the in main python process
    # No issues here
    generator = get_generator(seed)
else:
    # Multiprocess scenario. Throw an error if
    # we cannot access PyTorch's worker info.
    worker_info = torch.utils.data.get_worker_info()
    if worker_info is None:
        raise ValueError
    
    generator = get_generator(seed + worker_info.id)
```

Tests have also been written to verify this behaviour in `tests/trainer/test_data_collator.py`.

These changes were done on Python 3.12.8. The dependencies installed were as pip install -e ".[dev]" along with:

- torch==2.6.0
- tensorflow==2.18.0
- tf-keras==2.18.0

Fixes #36357 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@Rocketknight1 should be the right person to review this.

### CODE REVIEW

**Comment by capemox** [comment]
@Rocketknight1, this PR should be good to go given the fix above now

**Review by Rocketknight1 — APPROVED** [review]
On reading, this looks good to me - it adds a modest amount of code, but it's well-structured and everything is there for a reason.

The main thing I wanted to check was that there wouldn't be any regression if `self.seed = None`, but I believe that in this case `self.generator` will always be `None` as well, and therefore behaviour will be unchanged from before.

The only changes occur when `self.seed` is set, and previously this was unsupported at all, which means this PR is a strict improvement.

Still, this is big enough that we probably need core maintainer approval, so cc @arthurzucker @cyrilvallez, but since I own the data collator code you can just leave it to me to finalize the review and merge this if you want!

**Comment by capemox** [comment]
> The main thing I wanted to check was that there wouldn't be any regression if self.seed = None, but I believe that in this case self.generator will always be None as well, and therefore behaviour will be unchanged from before.

Yep, this is exactly the case! in the post_init method, self.generator is initialized to None, and there wouldn't be any changes compared to before. The `create_rng` method would initialize the generator, but that would only be the case if seed is set.

About the core maintainer review - since you're most familiar with the code, I'll leave it up to you! I wouldn't mind their inputs as well.

**Comment by Rocketknight1** [comment]
hi @capemox, it's up to the core maintainers to decide if they want to review this themselves or not!

**Comment by capemox** [comment]
Whoops nvm, thought it was directed at me :D

**Comment by Rocketknight1** [comment]
My bad lol, it did look like I was asking you to decide

**Review by Cyrilvallez — APPROVED** [review]
Alright, LGTM! About the tests, it looks like it's mostly copy/paste between frameworks, but as it seems to be the way other tests are as well, I'll let you judge @Rocketknight1! Feel free to merge when you're happy!

**Comment by Rocketknight1** [comment]
I'm happy with it at this point! Doing a rebase to hopefully clear the CI and then we can merge

**Comment by capemox** [comment]
My pleasure!
